drm: rp1: rp1-dpi: Add "rgb_order" property (to match VC4 DPI)

njhollinghurst opened the pull request:
As on VC4, the OF property overrides the order implied by media bus format. Only 4 of the 6 possible orders are supported. New add-on hardware designs should not rely on this "legacy" feature.

This is is an attempt to deal with #6505.

Not all combinations of DRM pixel format, bus format and rgb_order have been tested yet.

njhollinghurst commented:
I think this is correct, my only reticence (and the reason it's still a draft) is code size.
There might well be a smaller or cleaner way to achieve the permutation of input (shift,mask,scaling).
Alternatively, if we drop GRB, BRG support for `RGB565_1X16`, `RGB666_1X18` there might be a simpler version that permutes output shifts (though BGR666 will still need asymmetrical settings).

6by9 commented on drivers/gpu/drm/rp1/rp1-dpi/rp1_dpi.c:
```diff
@@ -353,6 +354,20 @@ static int rp1dpi_platform_probe(struct platform_device *pdev)
 	if (ret)
 		goto done_err;
 
+	dpi->rgb_order_override = -1;
```
My only comment is that this could potentially be defined as RP1DSPI_ORDER_UNCHANGED, and check for that by name instead of `< 0` at line 423.

njhollinghurst replied:
It's copied from vc4_dpi.c (though in that instance the values were from a register field).

BTW I also toyed with encoding the lookup order more directly in the variable (values like 0x020100) to avoid the switch statement.

6by9 replied:
vc4_dpi.c isn't always the best example, and there is no requirement to copy how it is implemented in the code, only the DT side.

njhollinghurst replied:
Added `RP1DSI_ORDER_UNCHANGED` as an explicit value, and re-worked the lookup to use bit-shifts (which makes both object and source code slightly smaller. That last `& 3` is superstitious since the values are internal).

njhollinghurst commented:
Tested on RGB666_CPADHI and RGB666 (Mode 5) HATs; kmstest with all supported 4CCs over all values of rgb_order.
The colours seemed to be going to the right places as far as I could see.

Also tested RGB565 (Mode 2) albeit without appropriate hardware, and the results looked plausible.

othermod commented:
Tested on my hardware as well. Verified rgb888 with rgb-order of brg and grb work as expected.

njhollinghurst committed:
drm: rp1: rp1-dpi: Add "rgb_order" property (to match VC4 DPI)

As on VC4, the OF property overrides the order implied by media
bus format. Only 4 of the 6 possible orders are supported. New
add-on hardware designs should not rely on this "legacy" feature.

Signed-off-by: Nick Hollinghurst <nick.hollinghurst@raspberrypi.com>
